### ISSUE TITLE
Login redirects to IP, not APP_URL

### ISSUE THREAD

**Issue by dspeckmann** [firefly-iii/firefly-iii#566]
I'm really enjoying this application so far, but there is one annoying issue I encountered. I am running Firefly in a docker container using nginx as a reverse proxy. I can reach it via `https://example.org/firefly`. Even though I've set `APP_FORCE_ROOT` and `APP_URL`, I get redirected to the IP of the container after I login, which obviously doesn't work. Once I manually enter the correct URL again everything works fine.

This is what the beginning of my `.env`  file looks like:

    APP_ENV=production
    APP_DEBUG=false
    APP_FORCE_SSL=true
    APP_FORCE_ROOT=https://example.org/firefly
    APP_KEY=xyz
    APP_LOG_LEVEL=warning
    APP_URL=https://example.org/firefly

Is there a best practive for a trailing backslash in `.env`? I've tried both and it doesn't change anything.

This is the simple nginx configuration:

    location /firefly/ {
        proxy_pass http://192.0.2.0:80/;
    }

IP and URL are different, of course.

**Comment by JC5** [comment]
I'll have to look into it. The internal redirect should indeed listen to the forced URL. But it's possible part of the application doesn't pick up on the forced url before it is sent out.

**Comment by JC5** [comment]
If you look at the source of your login page, what URL is the form referring to? 

```
<form action="https://firefly-iii.nder.be/login" method="post">
```

Is it the IP or the URL from `APP_FORCE_ROOT`?

**Comment by dspeckmann** [comment]
It points to the correct URL.

**Comment by JC5** [comment]
OK, I'm assuming you browse to the correct URL as well when you log in, which should take care of it. Locally, I cannot reproduce the problem 😞 . I have an installation on `firefly.app` and `firefly-alt-app/public/` and they work like a charm.

`APP_FORCE_ROOT=http://firefly-alt.app/public`

1. When I browse to the normal site, I get redirected to the login from alt. This login works as expected.
2. When I browse to the login of the normal site, I do not get redirected to the login from alt. However, I get a TokenMismatchException.

If you browse to the correct address (so not the IP address), _after_ you logged in (and it redirected you wrongly) are you in fact logged in?

If so, it might be that your Docker container is receiving the IP as referrer. Firefly III often uses the referrer field.

Try to disable the referrer field (this is different in many browsers). Or maybe it can be configured in nginx / Docker?

**Comment by dspeckmann** [comment]
Yes, it does log me in before redirecting.

The referrer field seems to be working and nothing changes when I disable it.

What is weird is that when I click the login button, then cancel and then click it again, it works. As far as I can see in Chrome the request headers are 100% identical, yet I get redirected to the IP the first time.

**Comment by dspeckmann** [comment]
Nginx (host) and Apache (container) access logs also show no differences between the two requests.

**Comment by JC5** [comment]
Do you cancel the login before it goes through? Has the page redirected you already? 

I have to dig deeper tomorrow.

**Comment by dspeckmann** [comment]
Yes, I cancel when the browser (doesn't matter which one) tries to access the IP.

I since changed the container from having a static IP to exposing the port, so now it's actually redirecting me to `http://localhost:8090`. This is also what's in the HTTP request's `Host` field, due to the reverse proxy. So it seems like after login I get redirected to the current host, not the referrer or the specified root URL.

**Comment by JC5** [comment]
I think I found it. See this commit:

https://github.com/firefly-iii/firefly-iii/commit/f0783df1239f2255ed4c7e2426f03ef8170bfc13

If you feel like it, you can make the change manually and see what happens.

**Comment by dspeckmann** [comment]
Unfortunately not, but I'm pretty sure I got it now. Before `redirectTo` is applied, Laravel tries to redirect you to your "intended URL" via `redirect()->intended(...)`. This method only seems to look at the Host header, not the Referrer. And in my case the host is `http://localhost:8090`. With those clues I finally found the `proxy_redirect` directive of nginx, which replaces the upstream server's address with the public one - Exactly what I needed! So in the end it was all my proxy's fault.

Thank you so much for the effort though!
